### PR TITLE
Improve global settings UI to be more intuitive/logical

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR improve global settings UI to be more intuitive/logical.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

### Screenshots (if appropriate):
Screenshot from 2023-01-05 16-16-29
Screenshot from 2023-01-05 16-18-22

### CODE REVIEW

**Comment by nvazquez** [comment]
@sureshanaparti is this PR still in progress?

**Comment by sureshanaparti** [comment]
> @sureshanaparti is this PR still in progress?

yes @nvazquez, work in progress.

**Comment by nvazquez** [comment]
@sureshanaparti can you please fix the conflicts?

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/MockConfigurationGroupVO.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+import org.apache.cloudstack.api.InternalIdentity;
+
+@Entity
+@Table(name = "mockconfigurationgroup")
```
Comment (nvazquez) [inline plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/MockConfigurationGroupVO.java:29]: Also here

File: plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/dao/MockConfigurationSubGroupDao.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// specific language governing permissions and limitations
+// under the License.
+package com.cloud.simulator.dao;
+
+import java.util.List;
+
+import com.cloud.simulator.MockConfigurationSubGroupVO;
+import com.cloud.utils.db.GenericDao;
+
+public interface MockConfigurationSubGroupDao extends GenericDao<MockConfigurationSubGroupVO, Long> {
```
Comment (nvazquez) [inline plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/dao/MockConfigurationSubGroupDao.java:24]: Same here, then your VO mock classes won't be needed

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -7311,6 +7319,82 @@ public String getConfigComponentName() {
...
+
+        if (type == String.class || type == Character.class) {
+            return Configuration.ValueType.String.name();
+        } else if (type == Integer.class || type == Long.class || type == Short.class) {
+            return Configuration.ValueType.Number.name();
+        } else if (type == Float.class || type == Double.class) {
+            return Configuration.ValueType.Decimal.name();
+        } else if (type == Boolean.class) {
+            return Configuration.ValueType.Boolean.name();
+        //} else if (type == Date.class) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:7364]: Can be removed

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41610to41700.java [hunk 4]
```diff
@@ -83,4 +92,99 @@ public void updateSystemVmTemplates(Connection conn) {
...
+                if (nameWords.length <= 0) {
+                    continue;
+                }
+
+                for (int index = 0; index < nameWords.length; index++) {
+                    Pair<Long, Long> configGroupAndSubGroup = getConfigurationGroupAndSubGroup(conn, nameWords[index]);
+                    if (configGroupAndSubGroup.first() != 1 && configGroupAndSubGroup.second() != 1) {
+                        stmt = "UPDATE `cloud`.`configuration` SET group_id = ?, subgroup_id = ? WHERE name = ?";
+                        pstmt = conn.prepareStatement(stmt);
+                        pstmt.setLong(1, configGroupAndSubGroup.first());
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41610to41700.java:120]: _configGroupAndSubGroup_  is not null at all.

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41610to41700.java [hunk 5]
```diff
@@ -83,4 +92,99 @@ public void updateSystemVmTemplates(Connection conn) {
...
+                            }
+                        }
+                    }
+                }
+                rsConfigurationSubGroups.close();
+            }
+            rs.close();
+            pstmt.close();
+        } catch (SQLException e) {
+            LOG.error("Failed to get configuration subgroup due to " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41610to41700.java:185]: set to default group / subgroup if any issues. throwing any error might block upgrade.

File: engine/schema/src/main/resources/META-INF/db/schema-41610to41700.sql [hunk 6]
```diff
@@ -655,3 +653,107 @@ INSERT INTO `cloud`.`user_vm_details`(`vm_id`, `name`, `value`)
...
+  `precedence` bigint(20) unsigned DEFAULT '999' COMMENT 'precedence for the configuration subgroup',
+  `group_id` bigint(20) unsigned NOT NULL COMMENT 'configuration group id',
+  PRIMARY KEY (`id`),
+  UNIQUE KEY (`name`, `group_id`),
+  CONSTRAINT `fk_configuration_subgroup__group_id` FOREIGN KEY (`group_id`) REFERENCES `configuration_group` (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+ALTER TABLE `cloud`.`configuration_group` AUTO_INCREMENT=1;
+
+INSERT INTO `cloud`.`configuration_group` (`name`, `description`, `precedence`) VALUES ('Miscellaneous', 'Miscellaneous configuration', 999);
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply engine/schema/src/main/resources/META-INF/db/schema-41610to41700.sql:683]: > Should we consider idempotent changes on these?

No (discussed offline)

File: plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/dao/MockConfigurationSubGroupDao.java (hunk 2 above)
Comment (sureshanaparti) [reply plugins/hypervisors/simulator/src/main/java/com/cloud/simulator/dao/MockConfigurationSubGroupDao.java]: removed the mock classes, not needed.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 7]
```diff
@@ -7311,6 +7319,82 @@ public String getConfigComponentName() {
...
+                    ConfigurationGroupVO configGroup = _configGroupDao.findById(configSubGroup.getGroupId());
+                    String groupName = configGroup != null ? configGroup.getName() : "Miscellaneous";
+                    return new Pair<String, String>(groupName, subGroupName);
+                }
+            } else {
+                s_logger.warn("Configuration " + configName + " not found");
+            }
+        }
+
+        s_logger.debug("Returning default configuration group for config: " + configName);
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:7386]: keeping default group for now, not to fail the response. will check.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
> Hi @ravening, your pull request has merge conflicts. Can you fix the conflicts and sync your branch with the base branch?

@sureshanaparti ^^

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by utchoang** [comment]
@blueorangutan ui

**Review by utchoang — COMMENTED** [review]
image
Please fix so that the margin of the breadcrumb is similar to that of other pages for reference below:
image

**Comment by sureshanaparti** [comment]
> image
> Please fix so that the margin of the breadcrumb is similar to that of other pages for reference below:
> image

@utchoang Fixed the margin

**Comment by utchoang** [comment]
@sureshanaparti With dark view, when hovering over the rows, it looks like below
image

File: ui/src/views/setting/AllConfigurationsTab.vue [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    margin-top: 20px;
+
+    @media (min-width: 760px) {
+      margin-top: 0;
+    }
+
+  }
+
+  .config-row-element {
+  margin-bottom: 10px;
```
Comment (utchoang) [inline ui/src/views/setting/AllConfigurationsTab.vue:270]:
```suggestion
    margin-bottom: 10px;
```

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan ui

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: ui/src/views/setting/AllConfigurationsTab.vue [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        size="small"
+        :pagination="false"
+        :loading="tabLoading"
+        :columns="columns"
+        :items="items"
+        :dataSource="items"
+        :columnKeys="columnKeys"
+        :rowKey="record => record.name"
+        :rowClassName="getRowClassName"
+        @refresh="this.fetchConfigData">
```
Comment (utchoang) [inline ui/src/views/setting/AllConfigurationsTab.vue:32]:
```suggestion
        @refresh="fetchConfigData">
```

File: ui/src/views/setting/AllConfigurationsTab.vue [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+      type: Boolean,
+      required: true
+    }
+  },
+  data () {
+    return {
+      apiName: 'listConfigurations',
+      columns: [
+        {
+          title: 'Name',
```
Comment (utchoang) [inline ui/src/views/setting/AllConfigurationsTab.vue:87]:
```suggestion
          title: this.$t('label.name'),
```

File: ui/src/views/setting/AllConfigurationsTab.vue [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+          dataIndex: 'name',
+          slots: { customRender: 'name' },
+          sorter: function (a, b) { return genericCompare(a[this.dataIndex] || '', b[this.dataIndex] || '') },
+          width: '60%'
+        },
+        {
+          title: 'Category',
```
Comment (utchoang) [inline ui/src/views/setting/AllConfigurationsTab.vue:94]:
```suggestion
          title: this.$t('label.category'),
```

File: ui/src/views/setting/AllConfigurationsTab.vue [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+          dataIndex: 'category',
+          slots: { customRender: 'category' },
+          sorter: function (a, b) { return genericCompare(a[this.dataIndex] || '', b[this.dataIndex] || '') }
+        },
+        {
+          title: 'Value',
```
Comment (utchoang) [inline ui/src/views/setting/AllConfigurationsTab.vue:100]:
```suggestion
          title: this.$t('label.value'),
```

File: ui/src/views/setting/ConfigurationTab.vue [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+<template>
+  <a-spin :spinning="tabLoading">
+    <a-table
+      size="small"
+      :showHeader="false"
+      :columns="columns"
+      :dataSource="this.configdata.filter(config => !config.parent)"
```
Comment (utchoang) [inline ui/src/views/setting/ConfigurationTab.vue:24]:
```suggestion
      :dataSource="configdata.filter(config => !config.parent)"
```

File: ui/src/views/setting/ConfigurationRow.vue [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+          <template #value="{ record }">
+            <ConfigurationValue :configrecord="record" :loading="rowLoading" :configDisabled="configDisabled" @change-config="onChangeConfig" />
+          </template>
+
+          <template #expandedRowRender="{}" v-if="childrenConfigData.length > 0">
+            <a-table
+              size="small"
+              :showHeader="false"
+              :columns="childrenColumns"
+              :dataSource="this.childrenConfigData"
```
Comment (utchoang) [inline ui/src/views/setting/ConfigurationRow.vue:42]:
```suggestion
              :dataSource="childrenConfigData"
```

File: ui/src/views/setting/ConfigurationTab.vue [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+      :rowKey="record => record.name"
+      :pagination="true"
+      :rowClassName="getRowClassName"
+      style="overflow-y: auto; margin-left: 10px" >
+
+      <template #displaytext="{ record }">
+        <ConfigurationRow :config="this.configdata" :configrecord="record" :loading="tabLoading" />
```
Comment (utchoang) [inline ui/src/views/setting/ConfigurationTab.vue:31]:
```suggestion
        <ConfigurationRow :config="configdata" :configrecord="record" :loading="tabLoading" />
```

File: ui/src/views/setting/GlobalConfiguration.vue [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+        </a-row>
+      </a-card>
+    </a-affix>
+    <a-tabs
+      tabPosition="left"
+      :animated="false"
+      @change="handleChangeConfigGroupTab" >
+      <a-tab-pane
+        key=''
+        tab='All Settings' >
```
Comment (utchoang) [inline ui/src/views/setting/GlobalConfiguration.vue:58]: Please add a new translation key

File: ui/src/views/setting/GlobalConfiguration.vue [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+      searchParams: {},
+      filter: '',
+      apiName: 'listConfigurations'
+    }
+  },
+  created () {
+    this.fetchConfigurationGroups()
+  },
+  watch: {
+  //   configGroup: {
```
Comment (utchoang) [inline ui/src/views/setting/GlobalConfiguration.vue:141]: Please remove these comments

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

File: ui/src/views/setting/ConfigurationRow.vue [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+      }
+      this.rowLoading = true
+      this.childrenConfigEnabled = (this.configrecord.value === 'true')
+      const params = {
+        parent: this.configrecord.name,
+        listAll: true
+      }
+      api('listConfigurations', params).then(response => {
+        this.childrenConfigData = response.listconfigurationsresponse.configuration
+        if (!this.childrenConfigData || this.childrenConfigData.length === 0) {
```
Comment (utchoang) [inline ui/src/views/setting/ConfigurationRow.vue:133]: I don't know what this paragraph is for. I think you can replace it like this `this.childrenConfigData = response?.listconfigurationsresponse?.configuration || []`

File: ui/src/views/setting/ConfigurationTab.vue [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  },
+  methods: {
+    fetchConfigurationDataByGroup (params = {}) {
+      this.tabLoading = true
+      params.pagesize = -1
+      console.log('group name: ' + this.group)
+      api('listConfigurations', params).then(response => {
+        this.configdata = response.listconfigurationsresponse.configuration
+        if (!this.configdata || this.configdata.length === 0) {
```
Comment (utchoang) [inline ui/src/views/setting/ConfigurationTab.vue:123]: Similar `this.configdata = response?.listconfigurationsresponse?.configuration || []`

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rohityadavcloud @nvazquez @utchoang @sureshanaparti is this ready?

**Comment by nvazquez** [comment]
@DaanHoogland there are still some minor bits to fix on this one

**Comment by borisstoyanov** [comment]
@blueorangutan ui

**Comment by borisstoyanov** [comment]
@blueorangutan test keepEnv

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan ui

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=5797)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=BUG) [2 Bugs](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=5797&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=5797&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=5797&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=CODE_SMELL) [77 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=5797&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=5797&metric=new_coverage&view=list) [18.4% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=5797&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=5797&metric=new_duplicated_lines_density&view=list) [2.5% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=5797&metric=new_duplicated_lines_density&view=list)

Can you look at these @davidjumani ?

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java [hunk 20]
```diff
@@ -136,14 +136,17 @@ public ConfigKey<?>[] getConfigKeys() {
         List<ConfigKey<String>> result = new ArrayList<ConfigKey<String>>();
 
         if (orderConfigKey != null && orderConfigKeyObj == null) {
-            orderConfigKeyObj = new ConfigKey<String>("Advanced", String.class, orderConfigKey, orderConfigDefault, "The order of precedence for the extensions", false);
+            orderConfigKeyObj = new ConfigKey<String>(String.class, orderConfigKey, "Advanced", orderConfigDefault, "The order of precedence for the extensions", false, ConfigKey.Scope.Global, null, null, null, null, null, ConfigKey.Kind.Order, orderConfigDefault);
```
Comment (DaanHoogland) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java:139]:
```suggestion
            orderConfigKeyObj = new ConfigKey<>(String.class, orderConfigKey, "Advanced", orderConfigDefault, "The order of precedence for the extensions", false, ConfigKey.Scope.Global, null, null, null, null, null, ConfigKey.Kind.Order, orderConfigDefault);
```

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java [hunk 21]
```diff
@@ -136,14 +136,17 @@ public ConfigKey<?>[] getConfigKeys() {
...
         }
 
+        // orderConfigKeyObj = new ConfigKey<String>("Advanced", String.class, orderConfigKey, orderConfigDefault, "The order of precedence for the extensions", false, Scope.Global, null, null, null, null, null, null, null);
+
+
```
Comment (DaanHoogland) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java:144]:
```suggestion
```
code in comment

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java [hunk 22]
```diff
@@ -136,14 +136,17 @@ public ConfigKey<?>[] getConfigKeys() {
...
         if (orderConfigKeyObj != null)
             result.add(orderConfigKeyObj);
 
         if (excludeKey != null && excludeKeyObj == null) {
-            excludeKeyObj = new ConfigKey<String>("Advanced", String.class, excludeKey, excludeDefault, "Extensions to exclude from being registered", false);
+            excludeKeyObj = new ConfigKey<String>(String.class, excludeKey, "Advanced", excludeDefault, "Extensions to exclude from being registered", false, ConfigKey.Scope.Global, null, null, null, null, null, ConfigKey.Kind.CSV, null);
```
Comment (DaanHoogland) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/registry/ExtensionRegistry.java:149]:
```suggestion
            excludeKeyObj = new ConfigKey<>(String.class, excludeKey, "Advanced", excludeDefault, "Extensions to exclude from being registered", false, ConfigKey.Scope.Global, null, null, null, null, null, ConfigKey.Kind.CSV, null);
```

File: plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/saml/SAML2AuthManager.java [hunk 23]
```diff
@@ -61,8 +61,8 @@ public interface SAML2AuthManager extends PluggableAPIAuthenticator, PluggableSe
     public static final ConfigKey<String> SAMLDefaultIdentityProviderId = new ConfigKey<String>("Advanced", String.class, "saml2.default.idpid", "https://openidp.feide.no",
             "The default IdP entity ID to use only in case of multiple IdPs", true);
 
-    public static final ConfigKey<String> SAMLSignatureAlgorithm = new ConfigKey<String>("Advanced", String.class, "saml2.sigalg", "SHA1",
-            "The algorithm to use to when signing a SAML request. Default is SHA1, allowed algorithms: SHA1, SHA256, SHA384, SHA512", true);
+    public static final ConfigKey<String> SAMLSignatureAlgorithm = new ConfigKey<String>(String.class, "saml2.sigalg", "Advanced", "SHA1",
```
Comment (DaanHoogland) [inline plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/saml/SAML2AuthManager.java:64]:
```suggestion
    public static final ConfigKey<String> SAMLSignatureAlgorithm = new ConfigKey<>(String.class, "saml2.sigalg", "Advanced", "SHA1",
```

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 24]
```diff
@@ -559,19 +563,58 @@ public ServiceOfferingResponse createServiceOfferingResponse(ServiceOffering off
...
     }
 
+    @Override
+    public ConfigurationGroupResponse createConfigurationGroupResponse(ConfigurationGroup cfgGroup) {
+        ConfigurationGroupResponse cfgGroupResponse = new ConfigurationGroupResponse();
+        cfgGroupResponse.setGroupName(cfgGroup.getName());
+        cfgGroupResponse.setDescription(cfgGroup.getDescription());
+        cfgGroupResponse.setPrecedence(cfgGroup.getPrecedence());
+
+        List<? extends ConfigurationSubGroup> subgroups = _configMgr.getConfigurationSubGroups(cfgGroup.getId());
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java:599]: if possible
```suggestion
        List<ConfigurationSubGroup> subgroups = _configMgr.getConfigurationSubGroups(cfgGroup.getId());
```

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java [hunk 25]
```diff
@@ -2214,6 +2250,20 @@ public Pair<List<? extends Configuration>, Integer> searchForConfigurations(fina
         return new Pair<List<? extends Configuration>, Integer>(result.first(), result.second());
     }
 
+    @Override
+    public Pair<List<? extends ConfigurationGroup>, Integer> listConfigurationGroups(ListCfgGroupsByCmd cmd) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/ManagementServerImpl.java:2254]:
```suggestion
    public Pair<List<ConfigurationGroup>, Integer> listConfigurationGroups(ListCfgGroupsByCmd cmd) {
```

?

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java [hunk 26]
```diff
@@ -2214,6 +2250,20 @@ public Pair<List<? extends Configuration>, Integer> searchForConfigurations(fina
...
+        final Filter searchFilter = new Filter(ConfigurationGroupVO.class, "precedence", true, null, null);
+        final SearchCriteria<ConfigurationGroupVO> sc = _configGroupDao.createSearchCriteria();
+
+        final String groupName = cmd.getGroupName();
+        if (StringUtils.isNotBlank(groupName)) {
+            sc.addAnd("name", SearchCriteria.Op.EQ, groupName);
+        }
+
+        final Pair<List<ConfigurationGroupVO>, Integer> result = _configGroupDao.searchAndCount(sc, searchFilter);
+        return new Pair<List<? extends ConfigurationGroup>, Integer>(result.first(), result.second());
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/ManagementServerImpl.java:2264]:
```suggestion
        return new Pair<List<ConfigurationGroup>, Integer>(result.first(), result.second());
```

File: test/integration/smoke/test_global_settings.py [hunk 27]
```diff
@@ -75,3 +75,120 @@ def tearDown(self):
...
+        cls.apiclient = cls.testClient.getApiClient()
+        cls._cleanup = []
+
+    @classmethod
+    def tearDownClass(cls):
+        try:
+            cleanup_resources(cls.apiclient, cls._cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+        return
```
Comment (DaanHoogland) [inline test/integration/smoke/test_global_settings.py:94]:
```suggestion
    def tearDownClass(cls):
        super(TestListConfigurations, cls).tearDownClass()
```

File: test/integration/smoke/test_global_settings.py [hunk 28]
```diff
@@ -75,3 +75,120 @@ def tearDown(self):
...
+
+    def tearDown(self):
+        """
+        Revert any configuration changes
+        """
+        try:
+            cleanup_resources(self.apiclient, self.cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+        return
```
Comment (DaanHoogland) [inline test/integration/smoke/test_global_settings.py:108]:
```suggestion
        super(TestListConfigurations, self).tearDown()
```

File: engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql [hunk 29]
```diff
@@ -993,3 +993,246 @@ BEGIN
     DECLARE CONTINUE HANDLER FOR 1061 BEGIN END; SET @ddl = CONCAT('ALTER TABLE ', in_table_name); SET @ddl = CONCAT(@ddl, ' ', ' ADD KEY ') ; SET @ddl = CONCAT(@ddl, ' ', in_index_name); SET @ddl = CONCAT(@ddl, ' ', in_key_definition); PREPARE stmt FROM @ddl; EXECUTE stmt; DEALLOCATE PREPARE stmt; END;
 
 CALL `cloud`.`IDEMPOTENT_ADD_KEY`('i_user_ip_address_state','user_ip_address', '(state)');
+--
+-- Update Configuration Groups and Subgroups
+--
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql:998]: none of the add columns below are idem potent

File: api/src/main/java/com/cloud/server/ManagementService.java [hunk 30]
```diff
@@ -102,6 +104,13 @@ public interface ManagementService {
      */
     Pair<List<? extends Configuration>, Integer> searchForConfigurations(ListCfgsByCmd c);
 
+    /**
+     * returns the the configuration groups
+     *
+     * @return list of configuration groups
+     */
+    Pair<List<? extends ConfigurationGroup>, Integer> listConfigurationGroups(ListCfgGroupsByCmd cmd);
```
[the remark this answers is not in the document]
Comment (davidjumani) [reply api/src/main/java/com/cloud/server/ManagementService.java:112]: Just for consistency across the function definitions in interface

File: engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql (hunk 29 above)
Comment (davidjumani) [reply engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql:998]: Adding a column in an idempotent way is not supported our of the box :(

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

File: engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql (hunk 29 above)
Comment (DaanHoogland) [reply engine/schema/src/main/resources/META-INF/db/schema-41720to41800.sql:998]: there is, `IDEMPOTENT_ADD_COLUMN` , but soit.

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Override
+    public ConfigurationSubGroupVO startsWithName(String name) {
+        SearchCriteria<ConfigurationSubGroupVO> sc = nameSearch.create();
+        sc.setParameters("name", name + "%");
+        return findOneIncludingRemovedBy(sc);
+    }
+
+    private ConfigurationSubGroupVO matchKeywordBy(BiPredicate<String, String> matcher, List<ConfigurationSubGroupVO> configurationSubGroups, String keyword) {
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java:75]: `matchBy` as a name would have said enough about the method

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Override
+    public ConfigurationSubGroupVO findByName(String name) {
+        SearchCriteria<ConfigurationSubGroupVO> sc = nameSearch.create();
+        sc.setParameters("name", name);
+        return findOneIncludingRemovedBy(sc);
+    }
+
+    @Override
+    public ConfigurationSubGroupVO startsWithName(String name) {
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java:69]: `startsWith` is descriptive enough

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+        nameAndGroupSearch.and("groupId", nameAndGroupSearch.entity().getGroupId(), SearchCriteria.Op.EQ);
+        nameAndGroupSearch.done();
+
+        keywordSearch = createSearchBuilder();
+        keywordSearch.and("keywords", keywordSearch.entity().getKeywords(), SearchCriteria.Op.NNULL);
+        keywordSearch.done();
+    }
+
+    @Override
+    public ConfigurationSubGroupVO findByName(String name) {
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java:62]: `findby`

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+                        return configurationSubGroup;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public ConfigurationSubGroupVO findByKeyword(String keyword) {
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/dao/ConfigurationSubGroupDaoImpl.java:100]: `findBy` again, too bad; `findAnywhere` as alternative

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test keepEnv

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test keepEnv

**Review by borisstoyanov — APPROVED** [review]
LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@davidjumani can you have a look at this conflict?

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Review by andrijapanicsb — APPROVED** [review]
LGTM, after manual testing and review.

**Pull request merged**